homeassistant/components/iqvia/__init__.py

```diff
@@ -251,7 +307,25 @@ def update():
             self.hass, TOPIC_DATA_UPDATE, update
```

Comment:
Instead of calling a callback which schedules an update, it's better to just do all the work directly and then call `self.async_write_ha_state()`.

Reply:
Not following. I use this pattern in all of my integrations. Can you expand?

Reply:
So currently your dispatcher callback is calling `async_schedule_update_ha_state(True)` which will start a state update in the next iteration of the event loop. The first step is to `await Entity.async_update()` which will finish, and then state writing can continue in the next iteration of the event loop.

None of these awaits are necessary and you should be able to extract all the state inside the dispatcher callback and then call `self.async_write_ha_state()`, which is a callback so you end up with 0 awaits. 

There are 2 approaches to implement this:
 1. Your properties will extract their values from the dispatcher's data.
 2. Your callback function will convert the data from the dispatcher to instance variables that your properties will return

Method 2 is necessary in case two properties require the same processing of the local state. 

For an example of 1, see https://github.com/home-assistant/core/blob/dev/homeassistant/components/updater/binary_sensor.py#L16

For an example of 2, see https://github.com/home-assistant/core/blob/dev/homeassistant/components/timer/__init__.py#L309-L318 . Note that in this case `async_finished` is actually incorrectly marked as async, fix here https://github.com/home-assistant/core/pull/32638.

Reply:
Got it; thanks for the explanation!

Reply:
7cafc9701d44

Reply:
7cafc9701d44